api/v1alpha1/proxydefaults_types_test.go

```diff
@@ -0,0 +1,313 @@
+package v1alpha1
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/hashicorp/consul-k8s/api/common"
+	capi "github.com/hashicorp/consul/api"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// Test MatchesConsul for cases that should return true.
+func TestProxyDefaults_MatchesConsulTrue(t *testing.T) {
+	cases := map[string]struct {
+		Ours   ProxyDefaults
+		Theirs *capi.ProxyConfigEntry
+	}{
+		"empty fields": {
+			Ours: ProxyDefaults{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: common.Global,
+				},
+				Spec: ProxyDefaultsSpec{},
+			},
+			Theirs: &capi.ProxyConfigEntry{
+				Name: common.Global,
+				Kind: capi.ProxyDefaults,
+			},
+		},
+		"all fields set": {
+			Ours: ProxyDefaults{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: common.Global,
+				},
+				Spec: ProxyDefaultsSpec{
+					Config: json.RawMessage(`{"envoy_tracing_json": "{\"http\":{\"name\":\"envoy.zipkin\",\"config\":{\"collector_cluster\":\"zipkin\",\"collector_endpoint\":\"/api/v1/spans\",\"shared_span_context\":false}}}"}`),
+					MeshGateway: MeshGatewayConfig{
+						Mode: "local",
+					},
+					Expose: ExposeConfig{
+						Checks: true,
+						Paths: []ExposePath{
+							{
+								ListenerPort:  80,
+								Path:          "/test/path",
+								LocalPathPort: 42,
+								Protocol:      "tcp",
+							},
+							{
+								ListenerPort:  8080,
+								Path:          "/root/test/path",
+								LocalPathPort: 4201,
+								Protocol:      "https",
+							},
+						},
+					},
+				},
+			},
+			Theirs: &capi.ProxyConfigEntry{
+				Kind: capi.ProxyDefaults,
+				Name: common.Global,
+				Config: map[string]interface{}{
+					"envoy_tracing_json": "{\"http\":{\"name\":\"envoy.zipkin\",\"config\":{\"collector_cluster\":\"zipkin\",\"collector_endpoint\":\"/api/v1/spans\",\"shared_span_context\":false}}}",
+				},
+				MeshGateway: capi.MeshGatewayConfig{
+					Mode: capi.MeshGatewayModeLocal,
+				},
+				Expose: capi.ExposeConfig{
+					Checks: true,
+					Paths: []capi.ExposePath{
+						{
+							ListenerPort:  80,
+							Path:          "/test/path",
+							LocalPathPort: 42,
+							Protocol:      "tcp",
+						},
+						{
+							ListenerPort:  8080,
+							Path:          "/root/test/path",
+							LocalPathPort: 4201,
+							Protocol:      "https",
+						},
+					},
+				},
+			},
+		},
+	}
+	for name, c := range cases {
+		t.Run(name, func(t *testing.T) {
+			require.True(t, c.Ours.MatchesConsul(c.Theirs))
+			require.Equal(t, c.Ours.ToConsul(), c.Theirs)
```

Comment:
Necessary?

Reply:
I stole this from one of your tests 😅

Reply:
shall drop

Reply:
![image](https://user-images.githubusercontent.com/1034429/94207228-0a464a80-fe7c-11ea-9391-add27dfc4793.png)

Which one? I should update it.

Reply:
https://github.com/hashicorp/consul-k8s/pull/335/files#diff-5881c5df1e9a27b333d30a2042ebcdeaR156-R158